roomedit: add support for disabling horizontal texture tiling

davidklaw opened the pull request:
## Description
This PR adds a new option flag to room sidedefs: No horizontal tile. When checked, the texture for a sidedef is clamped along the U axis rather than wrapped, which in some cases is causing a narrow errant slice of texture to be visible. This can be used to address issues like signs or webs, and other one-ended lines that aren't rendering as intended. This PR also adds clamping along the V axis when "No vertical tile" is applied.

You can learn more about texture clamping [here](https://learn.microsoft.com/en-us/windows/win32/direct3d9/clamp-texture-address-mode).

### Room Editor UI Changes
The editor's "LineDef Edit" and "Confirm copy of LineDef attributes" UI windows has been adjusted to accommodate this new flag.

### Future Work
There's still work to be done to update `.roo` files where this issue is present.

## Related Issues
#540

## Motivation and Context
There are one-sided linedefs throughout the world used for tavern signs, cob webs, and possibly more though I didn't do an extensive search. When in hardware rendering mode, the textures visibly wrap, ever so slightly. Clamping the sidedef cause the texture to be slightly stretched on one edge as opposed to show this edge wrap issue. See screenshots for examples.

## Screenshots:

davidklaw committed:
Create .clang-format

davidklaw committed:
Create .clang-format

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Revert "Create .clang-format"

This reverts commit 9019efef4b4e64942de7f8e3b7680360fae16bd2.

davidklaw committed:
Revert "Create .clang-format"

This reverts commit 9019efef4b4e64942de7f8e3b7680360fae16bd2.

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge branch 'master' of https://github.com/davidklaw/Meridian59

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge branch 'master' of https://github.com/davidklaw/Meridian59

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Merge remote-tracking branch 'upstream/master'

davidklaw committed:
Added support for a texture clamping flag

davidklaw committed:
Added support for a texture clamping flag

davidklaw committed:
Some formatting updates so as not to rock the boat

davidklaw committed:
Some formatting updates so as not to rock the boat

davidklaw committed:
Merge branch 'Meridian59:master' into texture-clamping

davidklaw committed:
Merge branch 'texture-clamping' of https://github.com/davidklaw/Meridian59 into texture-clamping

davidklaw commented:
I tried and failed to cleanup the commit history, so re-opening.

davidklaw committed:
Rolling back client change

davidklaw committed:
Updates layout and position for clamp flags

akirmse commented on roomedit/include/bsp.h:
```diff
@@ -56,6 +58,7 @@
 #define WF_BELOW_TOPDOWN  0x00000080      // Draw lower texture top-down
 #define WF_NORMAL_TOPDOWN 0x00000100      // Draw normal texture top-down
 #define WF_NO_VTILE       0x00000200      // Don't tile texture vertically (must be transparent)
+#define WF_CLAMP          0x00000400      // Clamp texture
```
Heads up that this value is not safe to use because 0x400 is already used for WF_HAS_ANIMATED in the client (whatever that is), in clientd3d/bsp.h.  Please copy over that client value so that this can be avoided in the future.

davidklaw committed:
Updated bit flags

davidklaw commented:
I'm at a point where I could use some support.

Where I'm at now, is that the Clamp flag can be enabled, but not then disabled. When debugging, I can see that the flag values remains the same when attempting to disable even though the UI check is seeing the unchecked checkbox.

I'm testing the flags at `lineedit.cpp:834`. I've been testing with `a1.roo` and the flags are initially 263172. When I remove 'No vertical tile' for the first sidedef, I get 1028 and then successfully revert it to 263172 by re-enabling it. When I enable clamp for the first sidedef, the flags update to 1311748 and remain unchanged when I try to disable clamp.

My best guess is that it has something to do with my bit flag choices, but I haven't been smart enough to figure out why 😐

davidklaw committed:
Update clamp bit flags

This commit updates the clamp flags to the only two remaining bit places I'm aware of and early tests have been successful using them.

davidklaw commented:
Utilizing the two high order bits seems to do the trick, though I need to do more extensive testing. Something feels icky about using the last remaining bits for these particular flags, but I guess that just means some future bridge that might need to be crossed. I don't see any cases of the flags being used in a signed integer context but I could use other eyes for assurance. Assuming there isn't, should I go around and updated the flags to unsigned ints? Seems pointless, but if it's best practice...

Meridian59 commented:
No need to go and change all the flag types.

On Mon, Dec 11, 2023 at 7:59 PM David Klawitter ***@***.***>
wrote:

> Utilizing the two high order bits seems to do the trick, though I need to
> do more extensive testing. Something feels icky about using the last
> remaining bits for these particular flags, but I guess that just means some
> future bridge that might need to be crossed. I don't see any cases of the
> flags being used in a signed integer context but I could use other eyes for
> assurance. Assuming there isn't, should I go around and updated the flags
> to unsigned ints? Seems pointless, but if it's best practice...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Meridian59/Meridian59/pull/583#issuecomment-1851260542>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARJPIJVFPUZ44GCBFGX2BDYI7JB7AVCNFSM6AAAAABAOTUTA2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNJRGI3DANJUGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

davidklaw committed:
Client clamping changes

davidklaw committed:
Roomedit clamp flag change

davidklaw committed:
New flag labeling, improved editor UI layout

davidklaw committed:
Updated app title and release build

davidklaw committed:
Applies texture mode in both U and V directions

davidklaw committed:
Revert "Applies texture mode in both U and V directions"

This reverts commit eb3bbe15b2683286f1b5cec4195da95977e9aa65.

davidklaw committed:
Refactors texture clamping into "No horizontal tile"

davidklaw committed:
Updates release build

davidklaw commented:
After discussing in Discord, it was decided to refactor this as a horizontal tile flag along the lines of the existing "No vertical tile" flag. When unchecked, the texture is wrapped. When checked, the texture is clamped.

akirmse commented on clientd3d/d3drender.c:
```diff
@@ -9280,6 +9292,16 @@ Bool D3DMaterialWorldStaticChunk(d3d_render_chunk_new *pChunk)
 		SetZBias(gpD3DDevice, ZBIAS_WORLD);
 	}
 
+	if (pChunk->flags & D3DRENDER_NO_VTILE)
```
In the spirit of repeating things less, I might write this instead as:

```
auto state = (pChunk->flags & D3DRENDER_NO_VTILE) ? D3DTADDRESS_CLAMP : D3DTADDRESS_WRAP;
IDirect3DDevice9_SetSamplerState(gpD3DDevice, 0, D3DSAMP_ADDRESSV, state);

```

Same thing for the other similar changes in this file.

davidklaw replied:
I'm not sure I love the way it turned out, though I did check in your suggestion. Setting it directly seemed more concise to me.

akirmse reviewed (commented):
Looks pretty clean, just one suggestion.

davidklaw committed:
Refactor logic

akirmse commented on clientd3d/d3drender.c:
```diff
@@ -9205,6 +9205,8 @@ Bool D3DMaterialWorldPacket(d3d_render_packet_new *pPacket, d3d_render_cache_sys
 
 Bool D3DMaterialWorldDynamicChunk(d3d_render_chunk_new *pChunk)
 {
+	auto state = NULL;
```
This is unexpected and leads to unnecessary if (state) checks later.  Try declaring state where it's first used.

davidklaw replied:
Got it, thanks.

davidklaw committed:
Optimizing the use of `state`
